### PR TITLE
DOC-229 | In-source detailed explanations of startup options 

### PR DESCRIPTION
**Pull request by Simran-B**
Forwardport of #17473 

Enterprise part: https://github.com/arangodb/enterprise/pull/1141

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -42,6 +42,11 @@ devel
   was useful with the MMFiles storage engine, but didn't have any useful
   effect when used with the RocksDB engine.
 
+* Added detailed explanations for some startup options.
+  They are only exposed via `--dump-options` under the `longDescription` key.
+
+* Improved shard distribution during collection creation.
```
Comment (jsteemann) [inline CHANGELOG:48]: Seems unrelated to the changes in this PR.

Comment (Simran-B) [inline CHANGELOG:48]:
```suggestion
```

File: arangod/Agency/AgencyFeature.cpp [hunk 2]
```diff
@@ -100,100 +100,110 @@ void AgencyFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
+                     "The step size between state machine compactions.",
                      new UInt64Parameter(&_compactionStepSize),
                      arangodb::options::makeFlags(
                          arangodb::options::Flags::DefaultNoComponents,
                          arangodb::options::Flags::Uncommon,
                          arangodb::options::Flags::OnAgent));
 
   options->addOption("--agency.compaction-keep-size",
-                     "keep as many indices before compaction point",
+                     "Keep as many indices before compaction point.",
```
Comment (jsteemann) [inline arangod/Agency/AgencyFeature.cpp:180]:
```suggestion
                     "Keep as many Agency log entries before compaction point.",
```

File: arangod/GeneralServer/GeneralServerFeature.cpp [hunk 3]
```diff
@@ -217,57 +217,82 @@ void GeneralServerFeature::collectOptions(
...
-  options->addOption("--http.keep-alive-timeout",
-                     "keep-alive timeout in seconds",
-                     new DoubleParameter(&_keepAliveTimeout));
+If the option is enabled and any of these headers is set, the request method is
+overridden by the value of the header. For example, this allows you to issue an
+HTTP DELETE request which, to the outside world, looks like an HTTP GET request.
+This allows you to bypass proxies and tools that only let certain request types
+pass.
+
+Enabling this option may impose a security risk. You should only used it in
```
Comment (jsteemann) [inline arangod/GeneralServer/GeneralServerFeature.cpp:239]:
```suggestion
Enabling this option may impose a security risk. You should only use it in
```

File: arangod/RestServer/EndpointFeature.cpp [hunk 4]
```diff
@@ -58,23 +58,132 @@ void EndpointFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
+GET / HTTP/1.1
+
+HTTP/1.1 301 Moved Permanently
+Location: /_db/_system/_admin/aardvark/index.html
+Content-Type: text/html
+Server: ArangoDB
+Connection: Keep-Alive
+Content-Length: 197
+
+<html><head><title>Moved</title></head><body><h1>Moved</h1><p>This page has moved to <a href="/_db/_system/_admin/aardvark/index.html">/_db/_system/_admin/aardvark/index.html</a>.</p></body></html>
```
Comment (jsteemann) [inline arangod/RestServer/EndpointFeature.cpp:154]: Seems superfluous here and too detailed. We shouldn't care too much about telnet's output in our documentation.

File: lib/ProgramOptions/Option.h [hunk 5]
```diff
@@ -137,6 +138,12 @@ struct Option {
   // format multiple version strings, comma-separated
   std::string toVersionString(std::vector<uint32_t> const& version) const;
 
+  // provide a detailed explanation of an option
+  Option& setLongDescription(char const* longDesc) {
```
Comment (jsteemann) [inline lib/ProgramOptions/Option.h:142]:
```suggestion
  Option& setLongDescription(std::string_view longDesc) {
```

File: lib/ProgramOptions/Option.h [hunk 6]
```diff
@@ -153,6 +160,9 @@ struct Option {
     return *this;
   }
 
+  // returns whether or not a long description was set
+  bool hasLongDescription() const { return !longDescription.empty(); }
```
Comment (jsteemann) [inline lib/ProgramOptions/Option.h:164]:
```suggestion
  bool hasLongDescription() const noexcept { return !longDescription.empty(); }
```

File: arangod/RestServer/EndpointFeature.cpp [hunk 7]
```diff
@@ -58,23 +58,132 @@ void EndpointFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
+```)");
```
Comment (Simran-B) [inline arangod/RestServer/EndpointFeature.cpp:155]:
```suggestion
You can use `telnet` to test the connection.)");
```

File: arangod/Transaction/ManagerFeature.cpp [hunk 8]
```diff
@@ -81,28 +81,30 @@ void ManagerFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
                   new DoubleParameter(&_streamingIdleTimeout, /*base*/ 1.0,
                                       /*minValue*/ 0.0,
                                       /*maxValue*/ maxStreamingIdleTimeout),
                   arangodb::options::makeFlags(
                       arangodb::options::Flags::DefaultNoComponents,
                       arangodb::options::Flags::OnCoordinator,
                       arangodb::options::Flags::OnSingle))
-      .setIntroducedIn(30800);
+      .setIntroducedIn(30800)
+      .setLongDescription(R"(Stream transactions automatically expire after
```
Comment (Simran-B) [inline arangod/Transaction/ManagerFeature.cpp:104]:
```suggestion
      .setLongDescription(R"(Stream transactions automatically expire after
```

**Review by jsteemann — APPROVED** [review]
LGTM

**Review by nerpaula — APPROVED** [review]
LGTM
